### PR TITLE
Fix location listener memory leak

### PR DESCRIPTION
**Pull request by thuutin**
### What is new in this Pull Request ? 
This PR fixes Location Listener memory leak

### CODE REVIEW

File: android-reactive-location/src/main/java/pl/charmas/android/reactivelocation/observables/BaseObservable.java [hunk 1]
```diff
@@ -30,7 +30,20 @@ protected BaseObservable(Context ctx, Api<? extends Api.ApiOptions.NotRequiredOp
...
+        for (Api<? extends Api.ApiOptions.NotRequiredOptions> service : services) {
+            apiClientBuilder.addApi(service);
+        }
+
+        apiClientBuilder.addConnectionCallbacks(apiClientConnectionCallbacks);
+        apiClientBuilder.addOnConnectionFailedListener(apiClientConnectionCallbacks);
+
+        final GoogleApiClient apiClient = apiClientBuilder.build();
+        apiClientConnectionCallbacks.setClient(apiClient);
+
```
Comment (thuytrinh) [inline android-reactive-location/src/main/java/pl/charmas/android/reactivelocation/observables/BaseObservable.java:46]: The setup isn't actually different from the deleted `createApiClient` right? If yes, I think we should leave it as it was.

File: android-reactive-location/src/main/java/pl/charmas/android/reactivelocation/observables/location/LocationUpdatesObservable.java [hunk 2]
```diff
@@ -44,6 +38,27 @@ protected void onUnsubscribed(GoogleApiClient locationClient) {
...
+        }
+
+        @Override
+        public void onLocationChanged(Location location) {
+            if (observer != null) {
+                observer.onNext(location);
+            }
+        }
+
+        void unSubscribe() {
```
Comment (thuytrinh) [inline android-reactive-location/src/main/java/pl/charmas/android/reactivelocation/observables/location/LocationUpdatesObservable.java]: Should be `unsubscribe()`?

File: android-reactive-location/src/main/java/pl/charmas/android/reactivelocation/observables/location/LocationUpdatesObservable.java [hunk 3]
```diff
@@ -44,6 +38,27 @@ protected void onUnsubscribed(GoogleApiClient locationClient) {
...
+            this.observer = null;
+        }
+
+
```
Comment (thuytrinh) [inline android-reactive-location/src/main/java/pl/charmas/android/reactivelocation/observables/location/LocationUpdatesObservable.java]: No empty line or only 1 empty line.

File: gradle.properties [hunk 4]
```diff
@@ -13,3 +13,4 @@ POM_LICENCE_URL=http://www.apache.org/licenses/LICENSE-2.0.txt
 POM_LICENCE_DIST=repo
 POM_DEVELOPER_ID=mcharmas
 POM_DEVELOPER_NAME=Michal Charmas
+REACTIVE_LOCATION_GMS_API_KEY
```
Comment (thuytrinh) [inline gradle.properties:16]: Why did you add this?

File: android-reactive-location/src/main/java/pl/charmas/android/reactivelocation/observables/location/LocationUpdatesObservable.java [hunk 5]
```diff
@@ -44,6 +38,27 @@ protected void onUnsubscribed(GoogleApiClient locationClient) {
         if (locationClient.isConnected()) {
             LocationServices.FusedLocationApi.removeLocationUpdates(locationClient, listener);
         }
+        listener.unSubscribe();
     }
 
+    static class UnSubcribeableLocationListener implements com.google.android.gms.location.LocationListener {
```
Comment (thuytrinh) [inline android-reactive-location/src/main/java/pl/charmas/android/reactivelocation/observables/location/LocationUpdatesObservable.java]: Should be `UnsubscribableLocationListener`?

Comment (thuytrinh) [reply android-reactive-location/src/main/java/pl/charmas/android/reactivelocation/observables/location/LocationUpdatesObservable.java]: > com.google.android.gms.location.LocationListener

Should import the package?

File: android-reactive-location/src/main/java/pl/charmas/android/reactivelocation/observables/BaseObservable.java (hunk 1 above)
Comment (thuutin) [reply android-reactive-location/src/main/java/pl/charmas/android/reactivelocation/observables/BaseObservable.java:46]: It is different. This setup allow me to keep a reference to the  `apiClientConnectionCallbacks` and unregister the callbacks when unsubscribed. 
``` 
subscriber.add(Subscriptions.create(new Action0() {
            @Override
            public void call() {
                if (apiClient.isConnected() || apiClient.isConnecting()) {
                    onUnsubscribed(apiClient);
                    apiClient.disconnect();
                    apiClient.unregisterConnectionFailedListener(apiClientConnectionCallbacks);
                    apiClient.unregisterConnectionCallbacks(apiClientConnectionCallbacks);
                }
            }
        }));

File: android-reactive-location/src/main/java/pl/charmas/android/reactivelocation/observables/location/LocationUpdatesObservable.java (hunk 2 above)
Comment (thuutin) [reply android-reactive-location/src/main/java/pl/charmas/android/reactivelocation/observables/location/LocationUpdatesObservable.java]: yes

**Comment by thuytrinh** [comment]
Got it.

File: gradle.properties (hunk 4 above)
Comment (thuutin) [reply gradle.properties:16]: For JitPack to build. It is because of the `build.gradle` file of the `samples` project.

File: android-reactive-location/src/main/java/pl/charmas/android/reactivelocation/observables/location/LocationUpdatesObservable.java (hunk 5 above)
Comment (thuutin) [reply android-reactive-location/src/main/java/pl/charmas/android/reactivelocation/observables/location/LocationUpdatesObservable.java]: Yes
